src/handlers/records-write.ts

```diff
@@ -146,77 +137,109 @@ export class RecordsWriteHandler implements MethodHandler {
+      isLatestBaseState = true;
+    } else {
+      const messageCid = await Message.getCid(message);
+      const result = await this.dataStore.put(tenant, messageCid, message.descriptor.dataCid, dataStream);
+      await this.validateDataStoreIntegrity(tenant, message, result.dataCid, result.dataSize);
+      isLatestBaseState = true;
+    }
+
+    return { isLatestBaseState, messageWithOptionalEncodedData };
+  }
+
+  private async processMessageWithoutDataStream(
+    tenant: string,
+    message: RecordsWriteMessage,
+    newestExistingWrite: RecordsWriteMessageWithOptionalEncodedData,
+  ):Promise<{
+    isLatestBaseState: boolean
+    messageWithOptionalEncodedData: RecordsWriteMessageWithOptionalEncodedData
+  }> {
+    let isLatestBaseState = false;
+    let messageWithOptionalEncodedData: RecordsWriteMessageWithOptionalEncodedData = message;
+    const { dataCid, dataSize } = message.descriptor;
+    // if the incoming message is not an initial write, and no dataStream is provided, first check integrity against newest existing write.
+    RecordsWriteHandler.validateDataIntegrity(dataCid, dataSize, newestExistingWrite.descriptor.dataCid, newestExistingWrite.descriptor.dataSize);
+    if (dataSize <= DwnConstant.maxDataSizeAllowedToBeEncoded) {
+      // we encode the data from the original write if it is smaller than the data-store threshold
+      if (newestExistingWrite.encodedData !== undefined) {
+        const dataBytes = Encoder.base64UrlToBytes(newestExistingWrite.encodedData);
+        messageWithOptionalEncodedData = await this.encodeAndSetData(message, dataBytes);
+        isLatestBaseState = true; // will show up in reads/queries
+      } else {
         throw new DwnError(
-          DwnErrorCode.RecordsWriteMissingDataInPrevious,
+          DwnErrorCode.RecordsWriteMissingEncodedDataInPrevious,
           `No dataStream was provided and unable to get data from previous message`
         );
-      } else {
-        dataBytes = Encoder.base64UrlToBytes(newestWithData.encodedData);
       }
     } else {
-      dataBytes = await DataStream.toBytes(dataStream);
+      const previousWriteMessageCid = await Message.getCid(newestExistingWrite);
+      // attempt to retrieve the data from the previous message, if it does not exist we have no previous data to associate.
+      // we preform this check in case a user attempts to gain access to data by knowing the dataCid,
+      // https://github.com/TBD54566975/dwn-sdk-js/issues/359
+      // so we insure that the data is already associated with the existing newest message
+      const dataResults = await this.dataStore.get(tenant, previousWriteMessageCid, message.descriptor.dataCid);
+      if (dataResults === undefined) {
+        throw new DwnError(
+          DwnErrorCode.RecordsWriteMissingDataInPrevious,
+          `No dataStream was provided and unable to get data from previous message`
+        );
+      }
```

Comment:
Sanity: I am wondering why would this condition be true in real life? The newest existing RecordsWrite should have the data right?